File: python/dataclasses.py
def _get_annotations(cls: type):
    inherits = []
    while cls is not object:
        inherits.append(cls)
        cls = cls.__base__
    inherits.reverse()
    res = {}
    for cls in inherits:
        res.update(cls.__annotations__)
    return res.keys()

def _wrapped__init__(self, *args, **kwargs):
    cls = type(self)
    cls_d = cls.__dict__
    fields = _get_annotations(cls)
    i = 0   # index into args
    for field in fields:
        if field in kwargs:
            setattr(self, field, kwargs.pop(field))
        else:
            if i < len(args):
                setattr(self, field, args[i])
                i += 1
            elif field in cls_d:    # has default value
                setattr(self, field, cls_d[field])
            else:
                raise TypeError(f"{cls.__name__} missing required argument {field!r}")
    if len(args) > i:
        raise TypeError(f"{cls.__name__} takes {len(fields)} positional arguments but {len(args)} were given")
    if len(kwargs) > 0:
        raise TypeError(f"{cls.__name__} got an unexpected keyword argument {next(iter(kwargs))!r}")

def _wrapped__repr__(self):
    fields = _get_annotations(type(self))
    obj_d = self.__dict__
    args: list = [f"{field}={obj_d[field]!r}" for field in fields]
    return f"{type(self).__name__}({', '.join(args)})"

def _wrapped__eq__(self, other):
    if type(self) is not type(other):
        return False
    fields = _get_annotations(type(self))
    for field in fields:
        if getattr(self, field) != getattr(other, field):
            return False
    return True

def _wrapped__ne__(self, other):
    return not self.__eq__(other)

def dataclass(cls: type):
    assert type(cls) is type
    cls_d = cls.__dict__
    if '__init__' not in cls_d:
        cls.__init__ = _wrapped__init__
    if '__repr__' not in cls_d:
        cls.__repr__ = _wrapped__repr__
    if '__eq__' not in cls_d:
        cls.__eq__ = _wrapped__eq__
    if '__ne__' not in cls_d:
        cls.__ne__ = _wrapped__ne__
    fields = _get_annotations(cls)
    has_default = False
    for field in fields:
        if field in cls_d:
            has_default = True
        else:
            if has_default:
                raise TypeError(f"non-default argument {field!r} follows default argument")
    return cls

def asdict(obj) -> dict:
    fields = _get_annotations(type(obj))
    obj_d = obj.__dict__
    return {field: obj_d[field] for field in fields}

File: src/dataclasses.cpp
#include "pocketpy/dataclasses.h"

namespace pkpy{

static void patch__init__(VM* vm, Type cls){
    vm->bind(vm->_t(cls), "__init__(self, *args, **kwargs)", [](VM* vm, ArgsView _view){
        PyObject* self = _view[0];
        const Tuple& args = CAST(Tuple&, _view[1]);
        const Dict& kwargs_ = CAST(Dict&, _view[2]);
        NameDict kwargs;
        kwargs_.apply([&](PyObject* k, PyObject* v){
            kwargs.set(CAST(Str&, k), v);
        });

        Type cls = vm->_tp(self);
        const PyTypeInfo* cls_info = &vm->_all_types[cls];
        NameDict& cls_d = cls_info->obj->attr();
        const auto& fields = cls_info->annotated_fields;

        int i = 0; // index into args
        for(StrName field: fields){
            if(kwargs.contains(field)){
                self->attr().set(field, kwargs[field]);
                kwargs.del(field);
            }else{
                if(i < args.size()){
                    self->attr().set(field, args[i]);
                    ++i;
                }else if(cls_d.contains(field)){    // has default value
                    self->attr().set(field, cls_d[field]);
                }else{
                    vm->TypeError(_S(cls_info->name, " missing required argument ", field.escape()));
                }
            }
        }
        if(args.size() > i){
            vm->TypeError(_S(cls_info->name, " takes ", fields.size(), " positional arguments but ", args.size(), " were given"));
        }
        if(kwargs.size() > 0){
            StrName unexpected_key = kwargs.items()[0].first;
            vm->TypeError(_S(cls_info->name, " got an unexpected keyword argument ", unexpected_key.escape()));
        }
        return vm->None;
    });
}

static void patch__repr__(VM* vm, Type cls){
    vm->bind__repr__(cls, [](VM* vm, PyObject* _0){
        auto _lock = vm->heap.gc_scope_lock();
        const PyTypeInfo* cls_info = &vm->_all_types[vm->_tp(_0)];
        const auto& fields = cls_info->annotated_fields;
        const NameDict& obj_d = _0->attr();
        SStream ss;
        ss << cls_info->name << "(";
        bool first = true;
        for(StrName field: fields){
            if(first) first = false;
            else ss << ", ";
            ss << field << "=" << CAST(Str&, vm->py_repr(obj_d[field]));
        }
        ss << ")";
        return VAR(ss.str());
    });
}

static void patch__eq__(VM* vm, Type cls){
    vm->bind__eq__(cls, [](VM* vm, PyObject* _0, PyObject* _1){
        if(vm->_tp(_0) != vm->_tp(_1)) return vm->NotImplemented;
        const PyTypeInfo* cls_info = &vm->_all_types[vm->_tp(_0)];
        const auto& fields = cls_info->annotated_fields;
        for(StrName field: fields){
            PyObject* lhs = _0->attr(field);
            PyObject* rhs = _1->attr(field);
            if(vm->py_ne(lhs, rhs)) return vm->False;
        }
        return vm->True;
    });
}

void add_module_dataclasses(VM* vm){
    PyObject* mod = vm->new_module("dataclasses");

    vm->bind_func<1>(mod, "dataclass", [](VM* vm, ArgsView args){
        vm->check_type(args[0], VM::tp_type);
        Type cls = PK_OBJ_GET(Type, args[0]);
        NameDict& cls_d = args[0]->attr();

        if(!cls_d.contains(__init__)) patch__init__(vm, cls);
        if(!cls_d.contains(__repr__)) patch__repr__(vm, cls);
        if(!cls_d.contains(__eq__)) patch__eq__(vm, cls);

        const auto& fields = vm->_all_types[cls].annotated_fields;
        bool has_default = false;
        for(StrName field: fields){
            if(cls_d.contains(field)){
                has_default = true;
            }else{
                if(has_default){
                    vm->TypeError(_S("non-default argument ", field.escape(), " follows default argument"));
                }
            }
        }
        return args[0];
    });

    vm->bind_func<1>(mod, "asdict", [](VM* vm, ArgsView args){
        const auto& fields = vm->_inst_type_info(args[0])->annotated_fields;
        const NameDict& obj_d = args[0]->attr();
        Dict d(vm);
        for(StrName field: fields){
            d.set(VAR(field.sv()), obj_d[field]);
        }
        return VAR(std::move(d));
    });
}

}   // namespace pkpy